image-diff

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/python3
+
+# This file is part of Cockpit.
+#
+# Copyright (C) 2020 Red Hat, Inc.
+#
+# Cockpit is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+
+from machine import testvm
+
+
+def get_packages(machine):
+    if any(machine.image.startswith(p) for p in ['debian', 'ubuntu']):
+        pkgcmd = "dpkg-query -W -f='PACKAGE\t${source:Package}\t${version}\n'"
+    elif any(machine.image.startswith(p) for p in ['rhel', 'fedora', 'centos']):
+        # SOURCERPM is not great and the others don't work
+        pkgcmd = "rpm -qa --qf 'PACKAGE\t%{NAME}\t%{EVR}\n'"
+    else:
+        raise ValueError("Can only list packages on rpm/dpkg, not " + machine.image)
+
+    output = machine.execute(pkgcmd)
+
+    packages = {}
+    for line in output.splitlines():
+        if line.startswith('PACKAGE'):
+            _, name, version = line.split('\t')
+            packages[name] = version
+
+    return packages
+
+
+parser = argparse.ArgumentParser(description='Compare CI images')
+parser.add_argument('old', help='the "old" image to compare')
+parser.add_argument('new', help='the "new" image to compare')
+args = parser.parse_args()
+
+# do everything in parallel
+old_vm = testvm.VirtMachine(image=args.old)
+new_vm = testvm.VirtMachine(image=args.new)
+
+old_vm.start()
+new_vm.start()
+
+old_vm.wait_boot()
+new_vm.wait_boot()
+
+old_pkgs = get_packages(old_vm)
+new_pkgs = get_packages(new_vm)
+
+old_vm.kill()
+new_vm.kill()
```

Comment:
DRY -- the VM handling could move to get_packages() or a new function, and then you just call it with the two images.

Reply:
I can't, though: I want to start the VMs in parallel and wait for them to boot in parallel, otherwise this takes twice as long.

I could probably move the `.wait_boot()` to inside `get_packages()` though.

Reply:
Ah, good point. So both the original and your new idea of moving just wait_boot() seem fine.

Reply:
I moved `wait_boot()` since it's a proper precondition of the `.execute()` in `get_packages()`, but `.kill()` feels too weird, so i'll leave it duplicated.

I briefly entertained putting both VMs into a list or a tuple and using some weird iterator notation to call methods on all items in the list but this just made things really unreadable.

Reply:
I take it back.  I actually prefer `.wait_boot()` also split out.  I've changed the comment above to emphasize that a bit more.